fix:  #128855 Ensure `Guard`'s `drop` method is removed at `opt-level=s` for `…

cblh opened the pull request:
fix: #128855

…Copy` types

Added `#[inline]` to the `drop` method in the `Guard` implementation to ensure that the method is removed by the compiler at optimization level `opt-level=s` for `Copy` types. This change aims to align the method's behavior with optimization expectations and ensure it does not affect performance.

r​? @scottmcm
<!-- homu-ignore:start -->
<!--
If this PR is related to an unstable feature or an otherwise tracked effort,
please link to the relevant tracking issue here. If you don't know of a related
tracking issue or there are none, feel free to ignore this.

This PR will get automatically assigned to a reviewer. In case you would like
a specific user to review your work, you can assign it to them by using

    r​? <reviewer name>
-->
<!-- homu-ignore:end -->

cblh committed:
fix: Ensure `Guard`'s `drop` method is removed at `opt-level=s` for `Copy` types

Added `#[inline]` to the `drop` method in the `Guard` implementation to ensure that the method is removed by the compiler at optimization level `opt-level=s` for `Copy` types. This change aims to align the method's behavior with optimization expectations and ensure it does not affect performance.

compiler-errors reviewed (changes requested):
This PR needs a test to actually demonstrate that it has the right effect.

tgross35 commented:
Looks like rustbot missed the assignment request

r​? @scottmcm

tgross35 commented:
Helloooooo rustbot

cblh committed:
test(std): Add codegen test for array::from_fn optimization

This commit adds a new test file 'array-from_fn.rs' to the codegen test suite.
The test checks the behavior of std::array::from_fn under different optimization levels:

1. At opt-level=0 (debug build), it verifies that the core::array::Guard
   is present in the generated code.
2. At opt-level=s (size optimization), it ensures that the Guard is
   optimized out.

This test helps ensure that the compiler correctly optimizes array::from_fn
calls in release builds while maintaining safety checks in debug builds.

bors commented:
:pushpin: Commit 3dc083dfd80d0436560c34c41be815b88510ffdc has been approved by `scottmcm`

It is now in the [queue](https://bors.rust-lang.org/homu/queue/rust) for this repository.

<!-- @bors r=scottmcm 3dc083dfd80d0436560c34c41be815b88510ffdc -->
<!-- homu: {"type":"Approved","sha":"3dc083dfd80d0436560c34c41be815b88510ffdc","approver":"scottmcm","queue":"https://bors.rust-lang.org/homu/queue/rust"} -->

bors commented:
:hourglass: Testing commit 3dc083dfd80d0436560c34c41be815b88510ffdc with merge 1d8f135b20fac63c493d5963ce02963b46ca0986...
<!-- homu: {"type":"BuildStarted","head_sha":"3dc083dfd80d0436560c34c41be815b88510ffdc","merge_sha":"1d8f135b20fac63c493d5963ce02963b46ca0986"} -->

bors commented:
:sunny: Test successful - [checks-actions](https://github.com/rust-lang-ci/rust/actions/runs/10346349998/job/28639037628)
Approved by: scottmcm
Pushing 1d8f135b20fac63c493d5963ce02963b46ca0986 to master...
<!-- homu: {"type":"BuildCompleted","approved_by":"scottmcm","base_ref":"master","builders":{"checks-actions":"https://github.com/rust-lang-ci/rust/actions/runs/10346349998/job/28639037628"},"merge_sha":"1d8f135b20fac63c493d5963ce02963b46ca0986"} -->

rust-timer commented:
Finished benchmarking commit (1d8f135b20fac63c493d5963ce02963b46ca0986): [comparison URL](https://perf.rust-lang.org/compare.html?start=13f8a57cfb57d527abe797230beabb64758bb888&end=1d8f135b20fac63c493d5963ce02963b46ca0986&stat=instructions:u).

### Overall result: ❌✅ regressions and improvements - no action needed


@rustbot label: -perf-regression



### [Instruction count](https://perf.rust-lang.org/compare.html?start=13f8a57cfb57d527abe797230beabb64758bb888&end=1d8f135b20fac63c493d5963ce02963b46ca0986&stat=instructions:u)
This is a highly reliable metric that was used to determine the overall result at the top of this comment.

|                                    | mean  | range          | count |
|:----------------------------------:|:-----:|:--------------:|:-----:|
| Regressions ❌ <br /> (primary)    | -     | -              | 0     |
| Regressions ❌ <br /> (secondary)  | 1.3%  | [1.3%, 1.3%]   | 1     |
| Improvements ✅ <br /> (primary)   | -     | -              | 0     |
| Improvements ✅ <br /> (secondary) | -1.0% | [-1.0%, -1.0%] | 1     |
| All ❌✅ (primary)                 | -     | -              | 0     |

### [Max RSS (memory usage)](https://perf.rust-lang.org/compare.html?start=13f8a57cfb57d527abe797230beabb64758bb888&end=1d8f135b20fac63c493d5963ce02963b46ca0986&stat=max-rss)
<details>
<summary>Results (primary 2.4%)</summary>

This is a less reliable metric that may be of interest but was not used to determine the overall result at the top of this comment.

|                                    | mean | range        | count |
|:----------------------------------:|:----:|:------------:|:-----:|
| Regressions ❌ <br /> (primary)    | 2.4% | [2.4%, 2.4%] | 1     |
| Regressions ❌ <br /> (secondary)  | -    | -            | 0     |
| Improvements ✅ <br /> (primary)   | -    | -            | 0     |
| Improvements ✅ <br /> (secondary) | -    | -            | 0     |
| All ❌✅ (primary)                 | 2.4% | [2.4%, 2.4%] | 1     |
</details>

### [Cycles](https://perf.rust-lang.org/compare.html?start=13f8a57cfb57d527abe797230beabb64758bb888&end=1d8f135b20fac63c493d5963ce02963b46ca0986&stat=cycles:u)
<details>
<summary>Results (secondary -5.6%)</summary>

This is a less reliable metric that may be of interest but was not used to determine the overall result at the top of this comment.

|                                    | mean  | range          | count |
|:----------------------------------:|:-----:|:--------------:|:-----:|
| Regressions ❌ <br /> (primary)    | -     | -              | 0     |
| Regressions ❌ <br /> (secondary)  | -     | -              | 0     |
| Improvements ✅ <br /> (primary)   | -     | -              | 0     |
| Improvements ✅ <br /> (secondary) | -5.6% | [-7.6%, -2.2%] | 3     |
| All ❌✅ (primary)                 | -     | -              | 0     |
</details>

### [Binary size](https://perf.rust-lang.org/compare.html?start=13f8a57cfb57d527abe797230beabb64758bb888&end=1d8f135b20fac63c493d5963ce02963b46ca0986&stat=size:linked_artifact)
This benchmark run did not return any relevant results for this metric.

**Bootstrap**: 752.51s -> 753.683s (0.16%)
**Artifact size**: 339.31 MiB -> 339.37 MiB (0.02%)
